In [31]:
!python -m ipykernel install --user --name=.venv

Installed kernelspec .venv in C:\Users\ayush\AppData\Roaming\jupyter\kernels\.venv


In [4]:
from tensorflow.keras.preprocessing.text import one_hot

In [6]:
sentences = [
    "The morning sun warmed the quiet street",
    "He placed the keys gently on the table",
    "The library was silent except for turning pages",
    "She tied her hair back before starting to paint",
    "The smell of fresh bread filled the kitchen",
    "Rainwater trickled down the old brick wall",
    "They watched the fireworks light up the sky",
    "The waves left seashells scattered along the shore",
    "A cat stretched lazily in the afternoon sun",
    "The train whistle echoed through the valley"
]


In [7]:
## define the vocab_size
vocab_size = 10000

In [9]:
### One hot representation
one_hot_repr = [one_hot(words, vocab_size) for words in sentences]
one_hot_repr

[[6275, 3626, 1651, 9447, 6275, 2577, 3865],
 [6719, 7764, 6275, 2609, 5563, 4793, 6275, 5456],
 [6275, 140, 7911, 3391, 4498, 5742, 5410, 6632],
 [8573, 7695, 7569, 3767, 7252, 4152, 4034, 1350, 1750],
 [6275, 845, 1648, 1181, 1738, 7038, 6275, 1001],
 [7743, 5072, 7346, 6275, 9324, 739, 6759],
 [7033, 1904, 6275, 4056, 3214, 7695, 6275, 9545],
 [6275, 540, 5912, 8915, 4217, 6951, 6275, 2507],
 [672, 827, 188, 3030, 3844, 6275, 1019, 1651],
 [6275, 590, 6236, 1885, 2236, 6275, 6153]]

## What does this mean?
### If vocab_size = 10000,
### for first word of sentence 1 : we have a vector [0, 0, 0, ..., 0, 1, 0, ..., 0]   # dimension = 10000
### The 1 is at position 6275. All other positions are 0.

In [39]:
from tensorflow.keras.layers import Embedding
from tensorflow.keras.utils import pad_sequences
from tensorflow.keras.models import Sequential
import tensorflow as tf

In [40]:
## Make all sentences of same length as RNN will take fixed number of timestamps
sent_len = 9
embedded_docs = pad_sequences(
  one_hot_repr,
  padding='pre', ## Add 0s at start
  maxlen=sent_len
)

embedded_docs ## All sentences of len=9

array([[   0,    0, 6275, 3626, 1651, 9447, 6275, 2577, 3865],
       [   0, 6719, 7764, 6275, 2609, 5563, 4793, 6275, 5456],
       [   0, 6275,  140, 7911, 3391, 4498, 5742, 5410, 6632],
       [8573, 7695, 7569, 3767, 7252, 4152, 4034, 1350, 1750],
       [   0, 6275,  845, 1648, 1181, 1738, 7038, 6275, 1001],
       [   0,    0, 7743, 5072, 7346, 6275, 9324,  739, 6759],
       [   0, 7033, 1904, 6275, 4056, 3214, 7695, 6275, 9545],
       [   0, 6275,  540, 5912, 8915, 4217, 6951, 6275, 2507],
       [   0,  672,  827,  188, 3030, 3844, 6275, 1019, 1651],
       [   0,    0, 6275,  590, 6236, 1885, 2236, 6275, 6153]])

In [41]:
## feature representation
dim=10

In [ ]:
model = Sequential()
model.add(Embedding(vocab_size, dim, input_length = sent_len)) ## internally it uses Word2Vec for embedding creation
model.compile(
  optimizer = tf.keras.optimizers.Adam(learning_rate=0.01),
  loss = tf.keras.losses.MSE
)

In [45]:
model.summary()

Model: "sequential_7"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_7 (Embedding)     (None, 9, 10)             100000    
                                                                 
Total params: 100000 (390.62 KB)
Trainable params: 100000 (390.62 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [47]:
model.predict(embedded_docs)

1/1 [==============================] - 0s 246ms/step


array([[[ 2.18277611e-02, -3.57678644e-02, -1.97297689e-02,
          9.69215482e-03, -8.54278728e-03,  3.57800387e-02,
          3.71795408e-02, -3.31172943e-02,  4.22801264e-02,
          3.86850946e-02],
        [ 2.18277611e-02, -3.57678644e-02, -1.97297689e-02,
          9.69215482e-03, -8.54278728e-03,  3.57800387e-02,
          3.71795408e-02, -3.31172943e-02,  4.22801264e-02,
          3.86850946e-02],
        [ 4.31544073e-02,  1.32532977e-02,  3.69069837e-02,
          2.52832100e-03,  8.18359852e-03,  2.88372375e-02,
         -4.92210500e-02, -1.91668868e-02,  3.27964872e-03,
          5.72804362e-03],
        [ 8.25200230e-03,  1.73392333e-02,  3.24430354e-02,
          1.35673992e-02, -6.53775781e-03, -2.13534962e-02,
         -3.42860445e-02, -3.55121717e-02,  4.15043496e-02,
         -2.07333807e-02],
        [-1.05254874e-02,  2.16152184e-02, -1.36049390e-02,
          3.29379551e-02,  9.71157476e-03,  1.43234469e-02,
         -2.29910854e-02, -2.02936418e-02,  2.236975

In [50]:
model.predict(embedded_docs).shape

1/1 [==============================] - 0s 43ms/step


(10, 9, 10)

### This means for 10 sentences -> each sentence has 9 words -> each word is a vector of dimension = 10

In [52]:
model.predict(embedded_docs)[0] ## first sentence

1/1 [==============================] - 0s 53ms/step


array([[ 0.02182776, -0.03576786, -0.01972977,  0.00969215, -0.00854279,
         0.03578004,  0.03717954, -0.03311729,  0.04228013,  0.03868509],
       [ 0.02182776, -0.03576786, -0.01972977,  0.00969215, -0.00854279,
         0.03578004,  0.03717954, -0.03311729,  0.04228013,  0.03868509],
       [ 0.04315441,  0.0132533 ,  0.03690698,  0.00252832,  0.0081836 ,
         0.02883724, -0.04922105, -0.01916689,  0.00327965,  0.00572804],
       [ 0.008252  ,  0.01733923,  0.03244304,  0.0135674 , -0.00653776,
        -0.0213535 , -0.03428604, -0.03551217,  0.04150435, -0.02073338],
       [-0.01052549,  0.02161522, -0.01360494,  0.03293796,  0.00971157,
         0.01432345, -0.02299109, -0.02029364,  0.02236975, -0.00913377],
       [ 0.00569329,  0.03682644,  0.02295345, -0.00256634, -0.04595074,
        -0.01538519,  0.04164454, -0.01895399, -0.0329292 ,  0.03786253],
       [ 0.04315441,  0.0132533 ,  0.03690698,  0.00252832,  0.0081836 ,
         0.02883724, -0.04922105, -0.01916689

In [55]:
model.predict(embedded_docs)[0][0] ## first word of first sentence

1/1 [==============================] - 0s 49ms/step


array([ 0.02182776, -0.03576786, -0.01972977,  0.00969215, -0.00854279,
        0.03578004,  0.03717954, -0.03311729,  0.04228013,  0.03868509],
      dtype=float32)